### PR TITLE
Fix error building image with docker compose

### PR DESCRIPTION
**Pull request by bu3tt1**
docker-compose build and docker-compose up give many errors. With this branch is possible to create the image and run it with docker-compose.

Modifications:
- .env file is not used, I removed it from docker-compose.yml
- generated asserts have not to be placed in mapped folders.

This resolve:
- https://github.com/laminas/getlaminas.org/issues/4
- https://github.com/laminas/getlaminas.org/issues/5

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,12 +43,16 @@ COPY templates /var/www/templates
 COPY config /var/www/config
 COPY src /var/www/src
 COPY data /var/www/data
+COPY public /var/www/public
+
+# Copy assets from docker assets stage in a not mapped folder to be maintained
+RUN mkdir -p /assets && mkdir -p /assets/css
+COPY --from=assets /assets/css/*.css /assets/css/
```
Comment (weierophinney) [inline Dockerfile]: I copied the assets in earlier, as they tend to change less frequently.

Additionally, I do not understand why you're not moving them to the final, production location, instead of an intermediary location that then requires copying them via the entrypoint script.

Can you elaborate on the rationale a bit more, please?

File: Dockerfile [hunk 2]
```diff
@@ -44,12 +43,16 @@ COPY templates /var/www/templates
...
 
 # Reset "local"/development config files
 RUN rm -f /var/www/config/development.config.php && \
-  rm /var/www/config/autoload/*.local.php && \
-  mv /var/www/config/autoload/local.php.dist /var/www/config/autoload/local.php
+    rm -f /var/www/config/autoload/*.local.php && \
+    mv /var/www/config/autoload/local.php.dist /var/www/config/autoload/local.php
```
Comment (weierophinney) [inline Dockerfile]: What prompted the indentation changes? Were you seeing errors? If so, can you provide:

- Docker and docker-compose version
- The error message(s) you observed

I don't see any myself when building images, so I'm curious about shi.

File: Dockerfile [hunk 3]
```diff
@@ -44,12 +43,16 @@ COPY templates /var/www/templates
...
 
 # Build project
 WORKDIR /var/www
-RUN composer install --quiet --no-ansi --no-dev --no-interaction --no-progress --no-scripts --no-plugins --optimize-autoloader
```
Comment (weierophinney) [inline Dockerfile]: Why is this moved to the entrypoint?

In production, this is no good, as it means it takes longer to deploy (we now have to wait for composer to install before the server can start).

File: bootstrap/gulpfile.js [hunk 4]
```diff
@@ -1,16 +1,18 @@
...
-  rename = require('gulp-rename'),
-  postcss      = require('gulp-postcss'),
-  autoprefixer = require('autoprefixer');
+var gulp = require('gulp');
+var sass = require('gulp-sass');
+var sourcemaps = require('gulp-sourcemaps');
+var cleanCss = require('gulp-clean-css');
+var rename = require('gulp-rename');
+var postcss = require('gulp-postcss');
+var autoprefixer = require('autoprefixer');
```
Comment (weierophinney) [inline bootstrap/gulpfile.js]: Are the above changes necessary? All versions of JS I've ever used allow grouping variable declarations in this way...

File: Dockerfile [hunk 5]
```diff
@@ -1,21 +1,24 @@
...
 FROM node:8-alpine as assets
 RUN apk add git
 RUN npm install -g gulp-cli
 RUN mkdir -p /work
 COPY bootstrap /work/
 WORKDIR /work
-RUN npm install && npm rebuild node-sass && gulp
+RUN npm install && npm rebuild node-sass && gulp sass
+# Created asset have to be moved in a not mapped folder to be maintained
+RUN mkdir -p /assets && mv /work/css /assets
```
Comment (weierophinney) [inline Dockerfile]: Why this change? If it's to have a semantically named directory, let's just create `/assets` instead of `/work`.

I'd also prefer we keep the `css` subdirectory, as we may be processing JS in the future as well, and doing so will make this image more future-proof.

File: Dockerfile [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
-# Build UI assets
+# Build UI assets stage
 FROM node:8-alpine as assets
 RUN apk add git
 RUN npm install -g gulp-cli
 RUN mkdir -p /work
 COPY bootstrap /work/
 WORKDIR /work
-RUN npm install && npm rebuild node-sass && gulp
+RUN npm install && npm rebuild node-sass && gulp sass
```
Comment (weierophinney) [inline Dockerfile]: Let's leave the `sass` target out, please; we may want to build more assets in the future, and I'd rather not have to hunt down why the assets are not present in the container.

File: Dockerfile (hunk 1 above)
Comment (bu3tt1) [reply Dockerfile]: When I use docker-compose.yml, if I don't put the files in an intermediate folder they are lost with the creation of the volume. This means that if I leave as it was previously, in the host folder, the generated assets were not found.

The alternative would be to run `gulp saas` via the entrypoint like `composer install`.

OS: Mac OS
Docker version 18.09.2
docker-compose version 1.23.2

File: Dockerfile (hunk 2 above)
Comment (bu3tt1) [reply Dockerfile]: There was no error related to the indentation, it was just to order the code.
The only problem was the lack of the parameter `-f` in the command `rm`.

OS: Mac OS
Docker version 18.09.2
docker-compose version 1.23.2

File: Dockerfile (hunk 3 above)
Comment (bu3tt1) [reply Dockerfile]: This was also done because the volume mapping lost the files generated by `composer install` with docker-compose.yml

Maybe there's another better alternative.

File: bootstrap/gulpfile.js (hunk 4 above)
Comment (bu3tt1) [reply bootstrap/gulpfile.js]: I don't think they're really necessary.

**Comment by bu3tt1** [comment]
The problems encountered occur the first time you create the image and instantiate the container. To replicate the errors you should delete the image from your host and empty the public/css folder as its content must be generated automatically.

My environment:
OS: Mac OS
Docker version 18.09.2
docker-compose version 1.23.2

File: Dockerfile (hunk 6 above)
Comment (bu3tt1) [reply Dockerfile]: If I remove it the image will not be created.

When I run: docker-compose up

[09:25:33] Using gulpfile /work/gulpfile.js
[09:25:33] Starting 'default'...
[09:25:33] Starting 'sass'...
[09:25:35] Finished 'sass' after 2.02 s
[09:25:35] Starting '<anonymous>'...

Then it waits and the script doesn't continue.

OS: Mac OS
Docker version 18.09.2
docker-compose version 1.23.2

File: Dockerfile (hunk 5 above)
Comment (bu3tt1) [reply Dockerfile]: If I use `work` that is a WORKDIR the generated assets are not copied in the other docker stage.

File: Dockerfile (hunk 3 above)
Comment (weierophinney) [reply Dockerfile]: Okay, I think I understand what's going on here finally, and it's a difference in development patterns.

When I'm developing locally, I'm running `composer` using a local PHP binary, so the `vendor` directory is present and usable at that point, as the application root directory is mapped to the `/var/www` directory in the container.

My guess is you're just firing up `docker-compose up` without doing a `composer install` first (or even later), which then means the entire application breaks for you.

I'm going to give the container a try within a bare checkout, and then see if the changes you propose fix them.

File: Dockerfile (hunk 6 above)
Comment (weierophinney) [reply Dockerfile]: I understand this change now... but don't know why it worked previously.

Basically, the `default` gulp target fires up `gulp.watch`, which is why that build step doesn't end for you.

Interestingly, in Linux, `gulp.watch` appears to detect it's in an environment that cannot spawn a browser, and it thus exits.

I think to solve this, I'm going to create a "deploy" target in the `gulpfile` where we can add multiple targets; that way, if we do build JS or other assets later, we can stuff them in there, and not need to change the `Dockerfile`.

File: Dockerfile (hunk 5 above)
Comment (weierophinney) [reply Dockerfile]: I think I understand why this change exists now, but there's clearly some differences in how Docker works between MacOS and Linux. On linux, the assets still get copied into the container volume, despite mapping the application root to `/var/www`.

File: Dockerfile (hunk 3 above)
Comment (weierophinney) [reply Dockerfile]: I have an idea on how to solve this; if it works, I'll push the changes to your branch.
